### PR TITLE
fix(server): update post updater iterator in tiering

### LINKED ISSUE #2488: Data tiering: fiber preemption and `PreUpdate && PostUpdate` might invalidate iterators
An example of this is `OpMutableGet` in `string_family.cc:179`. It first calls `auto res = op_args.shard->db_slice().FindAndFetchMutable(op_args.db_cntx, key);` which might preempt the fiber and interleave calls to`PreUpdate && PostUpdate`. Furthermore, iterators might be invalidated and a second Find is required. Also, investigate other commands as well.

### PR DESCRIPTION
**Pull request by adiholden**
fix #2488

### CODE REVIEW

File: src/server/db_slice.cc [hunk 1]
```diff
@@ -328,25 +328,32 @@ void DbSlice::AutoUpdater::Run() {
...
-
-  CHECK_EQ(fields_.deletion_count, fields_.db_slice->deletion_count_)
-      << "Attempting to run post-update after a deletion was issued";
-
   DCHECK(fields_.action == DestructorAction::kRun);
   CHECK_NE(fields_.db_slice, nullptr);
 
+  if (EngineShardSet::IsTieringEnabled()) {
+    // When triering is enabled we can preempt on write to disk, therefor it can be invalidated
+    // untill we run the post updated.
```
Comment (romange) [inline src/server/db_slice.cc:343]:
```suggestion
    // until we run the post updated.
```

File: src/server/db_slice.cc [hunk 2]
```diff
@@ -328,25 +328,32 @@ void DbSlice::AutoUpdater::Run() {
...
+    if (!fields_.it.IsOccupied() || fields_.it->first != fields_.key) {
```
Comment (romange) [inline src/server/db_slice.cc:344]: this check appears in several places now. Consider introducing an utility function:
`DbTable::Launder(PrimeIterator it, string_view key) -> PrimeIterator` that does that

File: src/server/engine_shard_set.cc [hunk 3]
```diff
@@ -836,7 +837,7 @@ void EngineShardSet::Init(uint32_t sz, bool update_db_time) {
                  << HumanReadableNumBytes(256_MB * shard_queue_.size()) << ". Exiting..";
       exit(1);
     }
-    is_tiering_enabled_ = true;
```
Comment (romange) [inline src/server/engine_shard_set.cc:839]: why did you change this?

File: src/server/db_slice.cc (hunk 1 above)
Comment (chakaz) [reply src/server/db_slice.cc:343]: therefor -> therefore

File: src/server/engine_shard_set.cc (hunk 3 above)
Comment (adiholden) [reply src/server/engine_shard_set.cc:839]: to be able to check is tiering is enabled without holding any object of engine shard set

Comment (romange) [reply src/server/engine_shard_set.cc:839]: shard_set is globally available object. why it matters?

Comment (adiholden) [reply src/server/engine_shard_set.cc]: you are right. I reverted this change

File: src/server/db_slice.cc [hunk 4]
```diff
@@ -328,25 +328,32 @@ void DbSlice::AutoUpdater::Run() {
...
+      fields_.it = fields_.db_slice->db_arr_[fields_.db_ind]->prime.Find(fields_.key);
+    }
+  } else {
+    // Make sure that the DB has not changed in size since this object was created.
+    // Adding or removing elements from the DB may invalidate iterators.
+    CHECK_EQ(fields_.db_size, fields_.db_slice->DbSize(fields_.db_ind))
+        << "Attempting to run post-update after DB was modified";
+
+    CHECK_EQ(fields_.deletion_count, fields_.db_slice->deletion_count_)
+        << "Attempting to run post-update after a deletion was issued";
```
Comment (chakaz) [inline src/server/db_slice.cc:354]: Can't we update these fields post-blocking so that we can keep these checks?

Comment (adiholden) [reply src/server/db_slice.cc:354]: we do the checks if we are not in tiering mode
If we are in tiering mode we write to disk at the end right before we do the post update so I dont think there is any value for updating the values and than checking them

Comment (chakaz) [reply src/server/db_slice.cc:354]: right, that makes sense

File: src/server/db_slice.cc (hunk 2 above)
Comment (chakaz) [reply src/server/db_slice.cc:344]: Do you not plan to do this?

Comment (adiholden) [reply src/server/db_slice.cc]: just added. Though I dont understand the function name suggested.. can you explain

Comment (chakaz) [reply src/server/db_slice.cc]: I think Vlad came up with that name, I'm not sure why launder specifically, but it sounds cool so I don't argue :sunglasses:

**Review by chakaz — APPROVED** [review]

**Pull request merged**